data/maps/SP_Scenarios/Along_The_River.wmf/scripting/functions.lua

```diff
@@ -0,0 +1,119 @@
+-- =======================================================================
+--                             Along The River
+-- =======================================================================
+
+-- Checks objective for done flag.
+function obj_done(objective)
+   return (objective ~= nil and objective.done)
+end
+
+-- =======================================================================
+
+-- Places building at field if player owns field and extra field for flag.
+function place_building_if_owner(player, building, point, cs)
+   if point.owner == player and point.brn.owner == player then
+      player:place_building(building, point, cs, true)
+      return true
+   end
+   return false
+end
+
+-- =======================================================================
+
+function remove_type(point, _type)
+   if point.immovable and point.immovable.descr.type_name == _type then
+      point.immovable:remove()
+   end
+end
+
+-- =======================================================================
+
+function immovable_has_name(immovable, name)
+   return (immovable ~= nil and immovable.descr.name == name)
+end
+
+-- =======================================================================
+
+function add_wares(immovable, ware, amount)
+   if immovable and (immovable.descr.type_name == "flag" or immovable.descr.type_name == "warehouse" or immovable.descr.type_name == "productionsite") then
+      immovable:set_wares(ware, immovable:get_wares(ware) + amount)
+   end
+end
+
+-- =======================================================================
+
+-- Transports wares to flag, warehouse or productionsite.
+function transport_wares(player, point, ware, amount, alert, image)
+   push_textdomain("map_along_the_river.wmf")
```

Comment:
This textdomain has a huge scope (it remains active during the lengthy while loop) but is needed only once below. Please move the `push_textdomain` call down, directly before the `if image then`, and the `pop_textdomain` to just after the corresponding `endif`.

--- 

The reason is that each Lua runtime has just 1 textdomain stack which is shared across all coroutines in the same environment. If a different coroutine pushes its own textdomain while this one here is sleeping, the whole stack will be corrupted. A push/pop pair of functions should therefore never, ever call `yield` (or its wrappers such as `sleep`) in-between so as to avoid this problem.